src/odemis/driver/autoscript_client.py

```diff
@@ -0,0 +1,2065 @@
+        return msg
+
+    def _updateSettings(self) -> None:
+        """
+        Reads all the current settings from the Detector and reflects them on the VAs
+        """
+        brightness = self.parent.get_brightness(self._scanner.channel)
+        if brightness != self.brightness.value:
+            self.brightness._value = brightness
+            self.brightness.notify(brightness)
+        contrast = self.parent.get_contrast(self._scanner.channel)
+        if contrast != self.contrast.value:
+            self.contrast._value = contrast
+            self.contrast.notify(contrast)
+        detector_type = self.parent.get_detector_type(self._scanner.channel)
+        if detector_type != self.type.value:
+            self.type._value = detector_type
+            self.type.notify(detector_type)
+        detector_mode = self.parent.get_detector_mode(self._scanner.channel)
+        if detector_mode != self.mode.value:
+            self.mode._value = detector_mode
+            self.mode.notify(detector_mode)
+
+    def _setBrightness(self, brightness: float) -> float:
+        self.parent.set_brightness(brightness, self._scanner.channel)
+        return self.parent.get_brightness(self._scanner.channel)
+
+    def _setContrast(self, contrast: float) -> float:
+        self.parent.set_contrast(contrast, self._scanner.channel)
+        return self.parent.get_contrast(self._scanner.channel)
+
+    def _setDetectorMode(self, mode: str) -> str:
+        self.parent.set_detector_mode(mode, self._scanner.channel)
+        return self.parent.get_detector_mode(self._scanner.channel)
+
+    def _setDetectorType(self, detector_type: str) -> str:
+        self.parent.set_detector_type(detector_type, self._scanner.channel)
+        return self.parent.get_detector_type(self._scanner.channel)
+
+    # TODO: add support for auto functions
+
+
+class TerminationRequested(Exception):
+    """
+    Generator termination requested.
+    """
+    pass
+
+
+class SEMDataFlow(model.DataFlow):
+    """
+    This is an extension of model.DataFlow. It receives notifications from the
+    detector component once the SEM output is captured. This is the dataflow to
+    which the SEM acquisition streams subscribe.
+    """
+
+    def __init__(self, detector):
+        """
+        detector (model.Detector): the detector that the dataflow corresponds to
+        """
+        model.DataFlow.__init__(self)
+        self._detector = detector
+
+    # start/stop_generate are _never_ called simultaneously (thread-safe)
+    def start_generate(self):
+        self._detector.start_generate()
+
+    def stop_generate(self):
+        self._detector.stop_generate()
+
+
+class Stage(model.Actuator):
+    """
+    This is an extension of the model.Actuator class. It provides functions for
+    moving the TFS stage and updating the position.
+    """
+
+    def __init__(self, name: str, role: str, parent: SEM, rng:Dict[str, Tuple[float]]=None, **kwargs):
```

Comment:
If `rng` accepts `None`, then it should be marked as "optional". Also, the Tuple type normally has one entry per element.
```suggestion
    def __init__(self, name: str, role: str, parent: SEM, rng: Optional[Dict[str, Tuple[float, float]]] = None, **kwargs):
```